server/src/main/java/org/elasticsearch/search/aggregations/InternalAggregation.java

```diff
@@ -127,6 +127,28 @@ public String getName() {
         return name;
     }
 
+    /**
+     * Rewrite the sub-aggregations in the buckets in this aggregation.
+     * The default implementation throws an exception because most
+     * aggregations don't <strong>have</strong> buckets in them. It
+     * should be overridden by aggregations that contain buckets.
+     */
+    public InternalAggregation rewriteBuckets(BucketRewriter rewriter) {
+        throw new IllegalStateException(
+                "Aggregation [" + getName() + "] must be a bucket aggregation but was [" + getWriteableName() + "]");
+    }
+    /**
+     * Used with {@link InternalAggregation#rewriteBuckets(BucketRewriter)} to
+     * rewrite the sub-aggregations within buckets.
+     */
+    public interface BucketRewriter {
```

Comment:
Does this gain us anything over just using `Function<InternalAggregations, InternalAggregations>`?  Shorter type signature, but it feels like it obfuscates what's going on?

Not a strong opinion on this, just felt like extra boilerplate.

Reply:
I thought it'd have other things but it turns out it doesn't. :+1: on the `Function`.